libkbfs/journal_md_ops.go

```diff
@@ -0,0 +1,382 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+)
+
+// A journalMDOps is an implementation of MDOps that delegates to a
+// TLF's mdJournal, if one exists. Specifically, it intercepts put
+// calls to write to the journal instead of the MDServer, where
+// something else is presumably flushing the journal to the MDServer.
+//
+// It then intercepts get calls to provide a combined view of the MDs
+// from the journal and the server when the journal is
+// non-empty. Specifically, if rev is the earliest revision in the
+// journal, and BID is the branch ID of the journal (which can only
+// have one), then any requests for revisions >= rev on BID will be
+// served from the journal instead of the server. If BID is empty,
+// i.e. the journal is holding merged revisions, then this means that
+// all merged revisions on the server from rev are hidden.
+//
+// TODO: This makes server updates meaningless for revisions >=
+// rev. Fix this.
+type journalMDOps struct {
+	MDOps
+	jServer  *JournalServer
+	mdServer MDServer
+}
+
+var _ MDOps = journalMDOps{}
+
+func (j journalMDOps) getHeadFromJournal(
+	ctx context.Context, id TlfID, bid BranchID, mStatus MergeStatus,
+	handle *TlfHandle) (
+	ImmutableRootMetadata, error) {
+	bundle, ok := j.jServer.getBundle(id)
+	if !ok {
+		return ImmutableRootMetadata{}, nil
+	}
+
+	_, uid, err := j.jServer.config.KBPKI().GetCurrentUserInfo(ctx)
+	if err != nil {
+		return ImmutableRootMetadata{}, err
+	}
+
+	bundle.lock.RLock()
+	defer bundle.lock.RUnlock()
+
+	head, err := bundle.mdJournal.getHead(uid)
+	if err != nil {
+		return ImmutableRootMetadata{}, err
+	}
+
+	if head == (ImmutableBareRootMetadata{}) {
+		return ImmutableRootMetadata{}, nil
+	}
+
+	if head.MergedStatus() != mStatus {
+		return ImmutableRootMetadata{}, nil
+	}
+
+	if mStatus == Unmerged && bid != NullBranchID && bid != head.BID {
+		return ImmutableRootMetadata{}, nil
+	}
+
+	if handle == nil {
+		bareHandle, err := head.MakeBareTlfHandle()
+		if err != nil {
+			return ImmutableRootMetadata{}, err
+		}
+		handle, err = MakeTlfHandle(
+			ctx, bareHandle, j.jServer.config.KBPKI())
+		if err != nil {
+			return ImmutableRootMetadata{}, err
+		}
+	}
+
+	rmd := RootMetadata{
+		BareRootMetadata: *head.BareRootMetadata,
+		tlfHandle:        handle,
```

Comment:
If a non-`nil` `handle` was passed in, should we be checking that it matches the MD?